gnome-image-installer/util/eos-reformatter.c

```diff
@@ -0,0 +1,1088 @@
+{
+  g_autoptr(GFileInfo) info = NULL;
+  g_autoptr(GFile) file = NULL;
+  GError *error = NULL;
+
+  file = g_file_new_for_path (reformatter->image);
+  info = g_file_query_info (file,
+                            G_FILE_ATTRIBUTE_STANDARD_CONTENT_TYPE ","
+                            G_FILE_ATTRIBUTE_STANDARD_SIZE,
+                            G_FILE_QUERY_INFO_NONE,
+                            NULL,
+                            &error);
+
+  if (info == NULL)
+    {
+      g_propagate_error(&reformatter->error, error);
+      g_error_free (error);
+    }
+  else
+    {
+      const gchar *type = NULL;
+
+      type = g_file_info_get_content_type (info);
+      if (g_str_has_suffix(type, "-xz-compressed"))
+        {
+          reformatter->decompressor = G_CONVERTER (gdu_xz_decompressor_new ());
+        }
+      else if (g_str_equal(type, "application/gzip"))
+        {
+          reformatter->decompressor = G_CONVERTER (g_zlib_decompressor_new (G_ZLIB_COMPRESSOR_FORMAT_GZIP));
+        }
+      else if (reformatter->write_total_bytes == 0)
+        {
+          reformatter->write_total_bytes = g_file_info_get_size (info);
+        }
+      if (reformatter->write_total_bytes == 0)
+        {
+          reformatter->error = g_error_new (EOS_REFORMATTER_ERROR,
+                                            EOS_REFORMATTER_ERROR_UNKNOWN_SIZE,
+                                            _("Target write size is unknown or unobtainable"));
+          return -1;
+        }
+  }
+
+  return open(reformatter->image, O_RDONLY);
+}
+
+static gboolean
+eos_reformatter_read (EosReformatter *reformatter, GAsyncQueue *freeq, GAsyncQueue *outq)
+{
+  gboolean ret = TRUE;
+  guint64 len = 0;
+  EosAlignedBuffer *buf = NULL;
+
+  if (g_cancellable_is_cancelled (reformatter->cancellable))
+    {
+      return FALSE;
+    }
+
+  buf = eos_reformatter_get_free_buffer (reformatter, freeq);
+
+  if (reformatter->gpg_in > 0)
+    {
+#ifdef USE_SPLICE
+      /* Splice to GPG stdin pipe, offering the offset so the read_fd position isn't changed */
+      len = splice (reformatter->read_fd, &reformatter->read_offset, reformatter->gpg_in, NULL, reformatter->buffer_size, 0);
+
+      if (len > 0)
+        {
+          /* Read to a buffer the same amount as splice */
+          buf->len = read (reformatter->read_fd, buf->ptr, len);
+          /* Update the current file position for next splice */
+          reformatter->read_offset = lseek (reformatter->read_fd, 0, SEEK_CUR);
+        }
+#else
+      buf->len = read (reformatter->read_fd, buf->ptr, reformatter->buffer_size);
+      if (buf->len > 0)
+        {
+          len = write (reformatter->gpg_in, buf->ptr, buf->len);
+        }
+#endif
+    }
+  else
+    {
+      /* Plain read, no GPG verification */
+      buf->len = read (reformatter->read_fd, buf->ptr, reformatter->buffer_size);
+    }
+
+  if (len < 0 || buf->len < 0)
+    {
+      g_mutex_lock (&reformatter->thread_mutex);
```

Comment:
This "lock; assign to `error`; unlock" logic is duplicated many times, it should be in a function. Also: every case fails to check whether `error` is non-`NULL`.